python/taichi/lang/kernel_impl.py

```diff
@@ -264,13 +264,19 @@ def extract_arg(arg, anno):
         if isinstance(anno, any_arr):
             if isinstance(arg, taichi.lang._ndarray.ScalarNdarray):
                 anno.check_element_dim(arg, 0)
+                anno.check_element_shapes(())
+                anno.check_field_dim(len(arg.shape))
                 return arg.dtype, len(arg.shape), (), Layout.AOS
             if isinstance(arg, taichi.lang.matrix.VectorNdarray):
                 anno.check_element_dim(arg, 1)
+                anno.check_element_shapes((arg.n, ))
+                anno.check_field_dim(len(arg.shape))
                 anno.check_layout(arg)
                 return arg.dtype, len(arg.shape) + 1, (arg.n, ), arg.layout
             if isinstance(arg, taichi.lang.matrix.MatrixNdarray):
                 anno.check_element_dim(arg, 2)
+                anno.check_element_shapes((arg.n, arg.m))
+                anno.check_field_dim(len(arg.shape))
```

Comment:
Some of the checks should be added to L284-294 after we have `element_shape` and `field_dim` parameters